util/types/json/json.go

```diff
@@ -15,26 +15,133 @@ package json
 
 import (
 	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unsafe"
 
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/hack"
 )
 
+const (
+	typeCodeObject  byte = 0x01
+	typeCodeArray   byte = 0x03
+	typeCodeLiteral byte = 0x04
+	typeCodeInt64   byte = 0x09
+	typeCodeFloat64 byte = 0x0b
+	typeCodeString  byte = 0x0c
+)
+
+const (
+	jsonLiteralNil   byte = 0x00
+	jsonLiteralTrue  byte = 0x01
+	jsonLiteralFalse byte = 0x02
+)
+
+const unknownTypeCodeErrorMsg = "unknown type code: %d"
+const unknownTypeErrorMsg = "unknown type: %s"
+
 // JSON is for MySQL JSON type.
-type JSON interface {
-	getTypeCode() byte
-	encode(buffer *bytes.Buffer)
-	// String returns JSON text.
-	String() string
-	// Type returns type string.
-	Type() string
+type JSON struct {
+	typeCode byte
+	i64      int64
+	str      string
+	object   map[string]JSON
+	array    []JSON
 }
 
 // CreateJSON creates a JSON from in. Return nil if in is not a valid JSON type.
 func CreateJSON(in interface{}) JSON {
 	return normalize(in)
 }
 
+// ParseFromString parses a json from string.
+func ParseFromString(s string) (j JSON, err error) {
+	// TODO: implement the decoder directly.
+	if len(s) == 0 {
+		err = ErrInvalidJSONText.GenByArgs("The document is empty")
+		return
+	}
+	if err = j.UnmarshalJSON(hack.Slice(s)); err != nil {
+		err = ErrInvalidJSONText.GenByArgs(err)
+	}
+	return
+}
+
+// MarshalJSON implements Marshaler interface.
+func (j JSON) MarshalJSON() ([]byte, error) {
+	switch j.typeCode {
+	case typeCodeObject:
+		return json.Marshal(j.object)
+	case typeCodeArray:
+		return json.Marshal(j.array)
+	case typeCodeLiteral:
+		switch byte(j.i64) {
+		case jsonLiteralNil:
+			return []byte("null"), nil
+		case jsonLiteralTrue:
+			return []byte("true"), nil
+		default:
+			return []byte("false"), nil
+		}
+	case typeCodeInt64:
+		return json.Marshal(j.i64)
+	case typeCodeFloat64:
+		f64 := *(*float64)(unsafe.Pointer(&j.i64))
+		return json.Marshal(f64)
+	case typeCodeString:
+		return json.Marshal(j.str)
+	default:
+		msg := fmt.Sprintf(unknownTypeCodeErrorMsg, j.typeCode)
+		panic(msg)
+	}
+}
+
+// UnmarshalJSON implements Unmarshaler interface.
+func (j *JSON) UnmarshalJSON(data []byte) (err error) {
+	var decoder = json.NewDecoder(bytes.NewReader(data))
+	decoder.UseNumber()
+	var in interface{}
+	if err = decoder.Decode(&in); err == nil {
+		*j = normalize(in)
+	}
+	return
+}
+
+// String implements fmt.Stringer interface.
+func (j JSON) String() string {
+	bytes, _ := json.Marshal(j)
+	return strings.TrimSpace(hack.String(bytes))
+}
+
+// Type returns type of JSON as string.
+func (j JSON) Type() string {
```

Comment:
I prefer use byte, because I need `buffer.WriteByte(typeCodeObject)`. if typedef them to other names, I need coerce them to byte.